juju/juju
# refactor(apiserver): move application and charm packages to model services

Refactor the following packages to use model info/config/agent services where possible, instead of model state methods:
- The application facade, `apiserver/facades/client/application`
- The charms facade, `apiserver/facades/client/charms`
- The charm downloader facade, `apiserver/facades/controller/charmdownloader`
- `internal/bootstrap`
- `internal/charm/services`
- `internal/worker/bootstrap`

These packages are all highly interlinked so it is easiest to do them all at once.

Unfortunately, we are not able to fully remove the `model` field from the application `APIBase`, as we still need the `OpenedPortRangesForMachine` method, and `stateenvirons.SupportedFeatures` is still using state model methods. 

## Checklist




## QA steps



The following should be sufficient to test all of the changed code:

Run the following test suites: `charmhub`, `deploy`, `deploy_caas`, `refresh`, `sidecar`.

Bootstrap Juju on k8s, deploy a simple application (e.g. `snappass-test`) and check that it deploys correctly.

Deploy an old v1 format charm, and check that it returns an error: `deploying format v1 charm "..." not supported`

Set the model config key `workload-storage` to an invalid value, deploy a charm with workload storage and check that it fails.
```
$ juju model-config workload-storage=foo
$ juju deploy mysql-k8s
$ juju status
App        Status  Scale  Charm      Channel      Message
mysql-k8s  error     0/1  mysql-k8s  8.0/stable   StorageClass.storage.k8s.io "m-foo" is invalid: provisioner: Required value
```

Check that you can `juju refresh` a charm with a LXD profile (e.g. `microk8s`).
```
# on a machine cloud
juju deploy microk8s --revision 200 --channel latest/edge
juju refresh microk8s
```

Set `charmhub-url` to a bogus value, and check that deploy fails.
```
$ juju add-model m --config charmhub-url=https://fake.com
$ juju deploy tiny-bash
ERROR resolving with preferred channel: unexpected charm-hub url "https://fake.com/v2/charms/refresh" when parsing headers
```

## Links



**Jira card:** JUJU-6218, JUJU-6379
